sw/host/opentitanlib/src/test_utils/elf_debugger.rs

```diff
@@ -0,0 +1,337 @@
+    pub fn write_u32(&self, addr: u32, value: u32) -> Result<()> {
+        self.write_memory32(addr, &[value])
+    }
+
+    /// Read the program counter.
+    pub fn get_pc(&self) -> Result<u32> {
+        self.read_riscv_reg(&RiscvReg::Csr(RiscvCsr::DPC))
+    }
+
+    /// Set the program counter to the given address.
+    pub fn set_pc(&self, address: impl Into<SymbolicAddress>) -> Result<()> {
+        let resolved = self.resolve(address)?;
+        log::info!("Set PC to {}", resolved);
+        self.write_reg(RiscvCsr::DPC, resolved.resolution)?;
+        Ok(())
+    }
+
+    /// Set a breakpoint at the given address.
+    pub fn set_breakpoint(&self, address: impl Into<SymbolicAddress>) -> Result<()> {
+        let resolved = self.resolve(address)?;
+        log::info!("Set breakpoint at {}", resolved);
+        self.jtag.set_breakpoint(resolved.resolution, true)?;
+        Ok(())
+    }
+
+    /// Assert the current program counter is at the given address.
+    pub fn expect_pc(&self, address: impl Into<SymbolicAddress>) -> Result<()> {
+        let resolved = self.resolve(address)?;
+        let pc = self.get_pc()?;
+        log::info!("PC = {:#x}, expected PC = {}", pc, resolved);
+        if pc != resolved.resolution {
+            anyhow::bail!("unexpected PC");
+        }
+        Ok(())
+    }
+
+    /// Assert the current program counter is within the given range.
+    pub fn expect_pc_range(&self, range: Range<impl Into<SymbolicAddress>>) -> Result<()> {
+        let start = self.resolve(range.start)?;
+        let end = self.resolve(range.end)?;
+        let pc = self.get_pc()?;
+        log::info!("PC = {:#x}, expected PC = {}..{}", pc, start, end,);
+        if !(start.resolution..end.resolution).contains(&pc) {
+            anyhow::bail!("unexpected PC");
+        }
+        Ok(())
+    }
+
+    /// Continue execution until the address is hit.
+    ///
+    /// Note that a breakpoint must not already exist for the target address, otherwise
+    /// this function will fail.
+    ///
+    /// This function will also fail if a pre-existing breakpoint is hit before the target
+    /// address is hit.
+    pub fn run_until(&self, address: impl Into<SymbolicAddress>, timeout: Duration) -> Result<()> {
+        let resolved = self.resolve(address)?;
+        log::info!("Run until {}", resolved);
+        self.jtag.set_breakpoint(resolved.resolution, true)?;
+        self.jtag.resume()?;
+        self.jtag.wait_halt(timeout)?;
+        self.expect_pc(resolved.resolution)?;
+        self.jtag.remove_breakpoint(resolved.resolution)?;
+        Ok(())
+    }
+
+    /// Excecute until the current function returns.
+    ///
+    /// This implementation does not use the debugging information from ELF files, and only use the RA register,
+    /// so it only works when RA has not been overriden, e.g. at the preamble of the function.
+    pub fn finish(&self, timeout: Duration) -> Result<()> {
```

Comment:
That's subjective but I think `finish_function` would have clearer than just `finish`.

Reply:
The corresponding gdb command is called `finish` so I follow the name.

Reply:
Ok, fair enough, but I don't find the GDB terminology particularly illuminating. For someone reading the code, I suspect this will force them to lookup the documentation of the function.